### ISSUE TITLE
cargo v0.27 tests failed when build rust 1.26.2 on "stable" channel

### ISSUE THREAD

**Issue by mnd** [rust-lang/cargo#5648]
If rust configured with 
```
[rust]
channel = "stable"
```
then test `test_resolving_minimum_version_with_transitive_deps` from `cargo/tests/testsuite/resolve.rs` will fail with next error:

```
---- resolve::test_resolving_minimum_version_with_transitive_deps stdout ----
        thread 'resolve::test_resolving_minimum_version_with_transitive_deps' panicked at 'called `Result::unwrap()` on an `Err` value: ErrorMessage { msg: "the `-Z` flag is only accepted on the nightly channel of Cargo" }', libcore/result.rs:945:5
```

**Comment by mnd** [comment]
Root cause: `config.configure` call tried to set `unstable_flags`, but `parse` function for `CliUnstable` structure allow it only on nightly channel.

**Comment by Eh2406** [comment]
Looks like that test is missing a `masquerade_as_nightly_cargo` or equivalent. Presumably we could just use [set_var](https://doc.rust-lang.org/std/env/fn.set_var.html) to do the equivalent.

I wonder why our ci did not complain about this. @mnd Where is that flag set?

**Comment by ehuss** [comment]
Due to the way `nightly_features_allowed` is written, it treats the unit-tests as the "dev" channel when `CFG_RELEASE_CHANNEL` environment var is not set (which is the case in CI).  When run from within the rust repo, it uses the channel defined in the env var, which in this case is "stable" which causes the test to fail.

I would be reluctant to use `set_var` without resetting it back to the original value afterwards.  I would think it would be easier just to skip the test if not running on nightly since it is a nightly-only feature.

**Comment by ehuss** [comment]
@mnd How are you running the tests.  When using `x.py test` in the rust repo it should set the `RUSTC_BOOTSTRAP` environment variable which allows this test to pass.

**Comment by mnd** [comment]
@ehuss I used https://static.rust-lang.org/dist/rustc-1.26.2-src.tar.gz sources,  defined `channel = "stable"` in `config.toml` file and started tests with `"./x.py" "-j1" "test" "src/tools/cargo"`

I have next environment variables defined: 
```
CFG_DISABLE_CROSS_TESTS=1
SHELL=/full/path/to/sh
CONFIG_SHELL=/full/path/to/sh
CC=/full/path/to/gcc
LLVM_LINK_SHARED=1
```

I preparing package for GuixSD and actually full config is [config.txt](https://github.com/rust-lang/cargo/files/2141129/config.txt), full environment is 
[environment-variables.txt](https://github.com/rust-lang/cargo/files/2141131/environment-variables.txt)

**Comment by ehuss** [comment]
Ah, I see.  I was looking at the wrong version.  The check for RUSTC_BOOSTRAP is very recent, and wasn't part of the 1.27 release.

If I understand the rust dist process correctly, tests are not run from `x.py dist`, so cargo is never tested with `channel="stable"`.

**Comment by Eh2406** [comment]
The PR to fix this was just approved. It insures that all currently existing tests that can't run on "stable" override the stability checks; In addition it makes it difficult to regress on this by assuming "stable" for tests unless explicitly overridden.

This fix will be in the 1.28 release of rust. I doubt they will backport these changes to past releases, but I have no say.
